llm-service/app/services/document_storage.py

```diff
@@ -0,0 +1,79 @@
+#  collective work is as provided in your written agreement with
+#  Cloudera.  Used apart from the collective work, this file is
+#  licensed for your use pursuant to the open source license
+#  identified above.
+#
+#  This code is provided to you pursuant a written agreement with
+#  (i) Cloudera, Inc. or (ii) a third-party authorized to distribute
+#  this code. If you do not have a written agreement with Cloudera nor
+#  with an authorized and properly licensed third party, you do not
+#  have any rights to access nor to use this code.
+#
+#  Absent a written agreement with Cloudera, Inc. ("Cloudera") to the
+#  contrary, A) CLOUDERA PROVIDES THIS CODE TO YOU WITHOUT WARRANTIES OF ANY
+#  KIND; (B) CLOUDERA DISCLAIMS ANY AND ALL EXPRESS AND IMPLIED
+#  WARRANTIES WITH RESPECT TO THIS CODE, INCLUDING BUT NOT LIMITED TO
+#  IMPLIED WARRANTIES OF TITLE, NON-INFRINGEMENT, MERCHANTABILITY AND
+#  FITNESS FOR A PARTICULAR PURPOSE; (C) CLOUDERA IS NOT LIABLE TO YOU,
+#  AND WILL NOT DEFEND, INDEMNIFY, NOR HOLD YOU HARMLESS FOR ANY CLAIMS
+#  ARISING FROM OR RELATED TO THE CODE; AND (D)WITH RESPECT TO YOUR EXERCISE
+#  OF ANY RIGHTS GRANTED TO YOU FOR THE CODE, CLOUDERA IS NOT LIABLE FOR ANY
+#  DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, PUNITIVE OR
+#  CONSEQUENTIAL DAMAGES INCLUDING, BUT NOT LIMITED TO, DAMAGES
+#  RELATED TO LOST REVENUE, LOST PROFITS, LOSS OF INCOME, LOSS OF
+#  BUSINESS ADVANTAGE OR UNAVAILABILITY, OR LOSS OR CORRUPTION OF
+#  DATA.
+#
+import os
+import shutil
+from abc import abstractmethod, ABC
+from pathlib import Path
+
+from . import s3
+from ..config import settings
+
+
+class DocumentStorage(ABC):
+    @abstractmethod
+    def download(self, temp_dir: str, bucket_name: str, document_key: str, original_filename: str) -> Path:
+        """
+        Copy file from storage into the temp directory
+        """
+
+    @staticmethod
+    def from_environment() -> "DocumentStorage":
+        # todo: move this to config, remove bucket_name from download api signature
+        bucket = os.environ.get("S3_RAG_DOCUMENT_BUCKET")
+        if bucket:
+            return S3DocumentStorage()
+        else:
+            return FileSystemDocumentStorage()
+
+class S3DocumentStorage(DocumentStorage):
+    def download(self, temp_dir: str, bucket_name: str, document_key: str, original_filename: str) -> Path:
+        """
+        Download document from S3
+        """
+        return s3.download(temp_dir, bucket_name, document_key, original_filename)
+
+
+class FileSystemDocumentStorage(DocumentStorage):
+    def download(self, temp_dir: str, bucket_name: str, document_key: str, original_filename: str) -> Path:
+        """
+        Copy file from local filesystem into the temp directory
+        """
+        source_file = Path(settings.rag_databases_dir, "file_storage", document_key)
+        target_file = Path(temp_dir, original_filename)
+        shutil.copy(source_file, target_file)
```

Comment:
nit: I think we can just return `source_file`. I don't think we depend on the specific file path for anything else anymore

Reply:
hmm. is that true? I thought the docling stuff needed the directory to write the temporary markdown file to

Reply:
oh yeah...we definitely need the temp version returned, since it has the original file name, unlike the saved one.